### PR TITLE
Better formatters

### LINKED ISSUE #238: feature request: `phrase hello slash world` -> `hello/world`
I'm not sure what the best way to go about this is, it might be simple.

some more examples/unit tests:

- `phrase hello slash world` -> `hello/world` (currently `hello /world`)
- `phrase hello slash snake name of file slash world` -> `hello/name_of_file/world` (currently `hello/name_of_file_/world`)

Similarly: `phrase file dot extension` -> `file.extension`, though this one might be complicated by the fact that I have a formatter called `dot` defined so that `dot hello world` -> `.hello_world` because I use python so much.

### PR DESCRIPTION
**Pull request by jcaw**
At dweil's request, a simple implementation of nested punctuation & chained (separate) formatters. 

(Details of each change in the commit messages)

Closes #238

### CODE REVIEW

**Comment by dwiel** [comment]
Looks good to me. Any thoughts from anyone else?

**Comment by knausj85** [comment]
I'm a little on the fence on including `user.numbers` in the capture, but I think it's fine to give whirl and tweak if needed.

**Comment by dwiel** [comment]
Yeah, what is the effect of the number capture here? "phrase the best one" -> "the best one" vs "the best1"?

**Comment by jcaw** [comment]
Number will be interpreted as a number - I can remove that. What about `"num <number>"` so the user can still enter numbers as e.g. part of a variable name?

**Comment by rntz** [comment]
I'm against including either <user.number> or <letter> without some qualifying prefix; otherwise I think too many ordinary words would get misinterpreted as letters or numbers.  I'm going to try out merging this branch on my local config and see if I run into any issues.

**Comment by rntz** [comment]
I have just tested this and it doesn't seem to work for me. For example, I tried to the phrase "snake object dot call method" and instead of `object.call_method` I get `object_dot_call_method`.

**Comment by dwiel** [comment]
This is exactly what unit tests are perfect for. Sure would be nice to have a simple test harness

**Comment by jcaw** [comment]
@rntz are you on Dragon or wav2lettter?

**Comment by rntz** [comment]
wav2letter.

**Comment by rntz** [comment]
@jcaw did you test this on Dragon, and did it work?

**Comment by jcaw** [comment]
Just to mirror from Slack: I've seen similar behaviour happening on Dragon. Sometimes dictation takes precedence, sometimes the command. I'm not sure what's happening - I'm using a chunked system similar to this on my personal config and this doesn't happen. Commands always take priority.

**Comment by rntz** [comment]
Further testing suggests that changing formatters does work but punctuation does not:

"hammer hello slash world" -> HelloSlashWorld :(
"snake hello world smash hello world" -> hello_worldhelloworld :)

**Comment by rntz** [comment]
whoops this was a mistake on my part, that happened because #240 is merged in my local config and renames <user.symbol> to <user.symbol_key>.

**Comment by rntz** [comment]
Planning to merge this as soon as I get around to it, probably in the next couple of days, unless someone beats me to it.

File: code/formatters.py [hunk 1]
```diff
@@ -228,15 +244,41 @@ def formatters_reformat_last(formatters: str) -> str:
...
 @ctx.capture(rule="{self.formatters}+")
 def formatters(m):
     return ",".join(m.formatters_list)
 
 
-@ctx.capture(rule="<self.formatters> <user.text>")
+@ctx.capture(
+    # Add anything else into this that you want to be able to speak during a
+    # formatter.
+    rule="(<user.symbol> | <user.letter> | <user.number>)"
```
Comment (rntz) [inline code/formatters.py:261]: This needs to be changed to use <user.number_key> and <user.symbol_key> now that #240 has been merged.

**Review by rntz — CHANGES_REQUESTED** [review]
The number/symbol vs number_key/symbol_key thing needs to be fixed before merging though. Otherwise everything looks good to me.

File: code/formatters.py (hunk 1 above)
Comment (jcaw) [reply code/formatters.py:261]: Ok - I'll gate "number" with a prefix too, that seems to be the consensus. I'll leave letters in for now, you guys experiment with it and gate that too if it becomes a problem.

**Comment by jcaw** [comment]
Should be compatible now.

**Pull request merged**